p2p/adapters/inproc.go

```diff
@@ -112,69 +180,99 @@ func (self *SimNode) setPeer(id *NodeId, rw *p2p.MsgPipeRW) *Peer {
 	return p
 }
 
-func (self *SimNode) Disconnect(rid []byte) error {
+func (self *SimNode) RemovePeer(node *discover.Node) {
 	self.lock.Lock()
 	defer self.lock.Unlock()
-	id := NewNodeId(rid)
+	id := &NodeId{node.ID}
 	peer := self.getPeer(id)
 	if peer == nil || peer.MsgPipeRW == nil {
-		return fmt.Errorf("already disconnected")
+		return
 	}
 	peer.MsgPipeRW.Close()
 	peer.MsgPipeRW = nil
 	// na := self.network.GetNodeAdapter(id)
 	// peer = na.(*SimNode).GetPeer(self.Id)
 	// peer.RW = nil
 	log.Trace(fmt.Sprintf("dropped peer %v", id))
-
-	return nil
 }
 
-func (self *SimNode) Connect(rid []byte) error {
+func (self *SimNode) AddPeer(node *discover.Node) {
 	self.lock.Lock()
 	defer self.lock.Unlock()
-	id := NewNodeId(rid)
+	id := &NodeId{node.ID}
 	na := self.network.GetNodeAdapter(id)
 	if na == nil {
-		return fmt.Errorf("node adapter for %v is missing", id)
+		panic(fmt.Sprintf("node adapter for %v is missing", id))
 	}
 	rw, rrw := p2p.MsgPipe()
 	// // run protocol on remote node with self as peer
 	peer := self.getPeer(id)
 	if peer != nil && peer.MsgPipeRW != nil {
-		return fmt.Errorf("already connected %v to peer %v", self.Id, id)
+		return
 	}
 	peer = self.setPeer(id, rrw)
 	close(peer.Connc)
 	defer close(peer.Readyc)
-	err := na.(ProtocolRunner).RunProtocol(self.Id, rrw, rw, peer)
-	if err != nil {
-		return fmt.Errorf("cannot run protocol (%v -> %v) %v", self.Id, id, err)
-	}
+	na.(*SimNode).RunProtocol(self, rrw, rw, peer)
 
 	// run protocol on remote node with self as peer
-	err = self.RunProtocol(id, rw, rrw, peer)
-	if err != nil {
-		return fmt.Errorf("cannot run protocol (%v -> %v): %v", id, self.Id, err)
-	}
+	self.RunProtocol(na.(*SimNode), rw, rrw, peer)
+}
+
+func (self *SimNode) PeerCount() int {
+	self.lock.Lock()
+	defer self.lock.Unlock()
+	return len(self.peers)
+}
+
+func (self *SimNode) NodeInfo() *p2p.NodeInfo {
+	return &p2p.NodeInfo{ID: self.Id.String()}
+}
+
+func (self *SimNode) PeersInfo() (info []*p2p.PeerInfo) {
 	return nil
 }
 
-func (self *SimNode) RunProtocol(id *NodeId, rw, rrw p2p.MsgReadWriter, peer *Peer) error {
-	if self.Run == nil {
+func (self *SimNode) RunProtocol(node *SimNode, rw, rrw p2p.MsgReadWriter, peer *Peer) {
+	id := node.Id
+	protocol := self.service.Protocols()[0]
+	if protocol.Run == nil {
 		log.Trace(fmt.Sprintf("no protocol starting on peer %v (connection with %v)", self.Id, id))
-		return nil
+		return
 	}
 	log.Trace(fmt.Sprintf("protocol starting on peer %v (connection with %v)", self.Id, id))
 	p := p2p.NewPeer(id.NodeID, id.Label(), []p2p.Cap{})
 	go func() {
 		self.network.DidConnect(self.Id, id)
```

Comment:
so we are still calling DIdConnect here. It is ok
For other types we need to implement a subscription via IPC/WS to signal the connection events.
In which case inproc can just behave the same, analogous to connect (via admin_addPeer)